src/compiler/checker.ts

```diff
@@ -25500,17 +25500,32 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             if (isKnownProperty) {
                 // If the check is for a known property (i.e. a property declared in some constituent of
                 // the target type), we filter the target type by presence of absence of the property.
-                return filterType(type, t => isTypePresencePossible(t, name, assumeTrue));
+                const filtered = filterType(type, t => isTypePresencePossible(t, name, assumeTrue));
+                // without `noUncheckedIndexedAccess` we can't narrow any further
+                // `{ [x: PropertyKey]: T }[prop]` already returns `T`
+                if (!compilerOptions.noUncheckedIndexedAccess) {
+                    return filtered;
+                }
+                return mapType(filtered, t => {

```

Comment:
I think you should only do this if `assumeTrue` is true (maybe have the check above be `if (!compilerOptions.noUncheckedIndexedAccess || !assumeTrue) ...`).

Might be nice to add that to the tests as well, something like:
```ts
function f2(obj: Record<string, string>) {
  if ("test" in obj) {
    return obj.test
  }
  else {
    obj.test // Here it should be `string | undefined` still
  }
  return "default"
}
```

Reply:
great catch! thanks. I've pushed out the fix for this